src/snowflake/snowpark/_internal/ast_utils.py

```diff
@@ -386,53 +386,34 @@ def build_udtf_apply(
     build_fn_apply_args(ast, *args, **kwargs)
 
 
-def build_session_table_fn_apply(
+def build_call_table_function_apply(
     ast: proto.Expr,
-    name: Union[str, Iterable[str]],
+    name: str,
     *args: Tuple[Union[proto.Expr, Any]],
     **kwargs: Dict[str, Union[proto.Expr, Any]],
 ) -> None:
-    """
-    Creates AST encoding for ApplyExpr(SessionTableFn(<name>), List(<args...>), Map(<kwargs...>)) for session table functions.
-    Args:
-        ast: Expr node to fill.
-        name: Name of the session table function to call.
-        *args: Positional arguments to pass to function.
-        **kwargs: Keyword arguments to pass to function.
-    """
     expr = with_src_position(ast.apply_expr)
-    _set_fn_name(name, expr.fn.session_table_fn)
-    with_src_position(expr.fn.session_table_fn)
+    _set_fn_name(name, expr.fn.call_table_function_expr)
+    with_src_position(expr.fn.call_table_function_expr)
     build_fn_apply_args(ast, *args, **kwargs)
 
 
-def build_table_fn_apply(
+def build_indirect_table_fn_apply(
     ast: proto.Expr,
-    name: Union[str, Iterable[str], None],
-    *args: Tuple[Union[proto.Expr, Any]],
-    **kwargs: Dict[str, Union[proto.Expr, Any]],
+    func: Union[str, List[str], "snowflake.snowpark.table_function.TableFunctionCall"],
+    *func_arguments: ColumnOrName,
+    **func_named_arguments: ColumnOrName,
 ) -> None:
-    """
-    Creates AST encoding for ApplyExpr(TableFn(<name>), List(<args...>), Map(<kwargs...>)) for table functions.
-    Args:
-        ast: Expr node to fill.
-        name: Name of the table function to call. The name can be None and is ignored for table function calls of type SessionTableFn.
-        *args: Positional arguments to pass to function.
-        **kwargs: Keyword arguments to pass to function.
-
-    Requires that ast.apply_expr.fn.table_fn.call_type is set to a valid TableFnCallType.
-    """
     expr = with_src_position(ast.apply_expr)
-    assert (
-        ast.apply_expr.fn.table_fn.call_type.WhichOneof("variant") is not None
-    ), f"Explicitly set the call type before calling this function {str(ast.apply_expr.fn.table_fn)}"
-    if not expr.fn.table_fn.call_type.table_fn_call_type__session_table_fn:
-        assert (
-            name is not None
-        ), f"Table function name must be provided {str(ast.apply_expr.fn.table_fn)}"
-        _set_fn_name(name, expr.fn.table_fn)
-    with_src_position(expr.fn.table_fn)
-    build_fn_apply_args(ast, *args, **kwargs)
+    if hasattr(func, "_ast_stmt"):
```

Comment:
I would actually prefer to have this more explicit, i.e. no indirect attribute lookup but rather:
```
if isinstance(func, TableFunctionCall):
   stmt = func._ast_stmt.
```
that is much easier to read. You can use local imports to prevent circular import issues. Much preferred to the `hasattr` solution.

Reply:
Multiple types are involved, though. `Callable` objects created from `functions.table_function` also have this property.

Which one do you prefer?

1. Explicit type enumeration (`if isinstance(func, (TableFunctionCall, Callable)):`)
2. Just look up the desired field

Reply:
I think 1. would be easier, thinking here of maintenance. If we ever wanted to support a new entity, very clear what to do/which branch logic takes. Looking up a potentially existing field seems more confusing imho.

Reply:
Done